### PR TITLE
fix nill state timeouts and enhance v08 test

### PR DESCRIPTION
**Pull request by spolti**
Signed-off-by: spolti <filippespolti@gmail.com>

**Many thanks for submitting your Pull Request :heart:!**

**What this PR does / why we need it**:

**Special notes for reviewers**:

**Additional information (if needed):**

### CODE REVIEW

File: parser/parser_test.go [hunk 1]
```diff
@@ -105,6 +90,7 @@ func TestFromFile(t *testing.T) {
 				assert.Equal(t, "CheckVisaStatus", w.States[1].GetName())
 				assert.IsType(t, &model.DataBasedSwitchState{}, w.States[0])
 				assert.IsType(t, &model.EventBasedSwitchState{}, w.States[1])
+				//assert.Equal(t, "PT1H", w.States[1].(*model.EventBasedSwitchState).Timeouts.EventTimeout)
```
Comment (ricardozanini) [inline parser/parser_test.go:93]: Why we commented this one? If it's not needed, we can delete.

File: parser/parser_test.go [hunk 2]
```diff
@@ -82,21 +82,6 @@ func TestFromFile(t *testing.T) {
...
-				assert.Equal(t, "custom.greeting", w.ID)
-				assert.NotEmpty(t, w.States[0].(*model.OperationState).Actions)
-				assert.NotEmpty(t, w.Functions[0])
-				assert.Equal(t, "greetingCustomFunction", w.Functions[0].Name)
-				assert.Equal(t, model.FunctionTypeCustom, w.Functions[0].Type)
-				assert.NotNil(t, w.States[0].(*model.OperationState).Actions[0].FunctionRef)
-				assert.Equal(t, "greetingCustomFunction", w.States[0].(*model.OperationState).Actions[0].FunctionRef.RefName)
-				assert.Equal(t, "greetingCustomFunction", w.States[0].(*model.OperationState).Actions[0].Name)
-			},
 		}, {
```
Comment (ricardozanini) [inline parser/parser_test.go:85]: This test is not used anymore?

Comment (spolti) [reply parser/parser_test.go:85]: I moved it to the last item in this array and added more checks there.

**Review by ricardozanini — COMMENTED** [review]
LGTM! Many thanks! @lsytj0413 wanna take a look?

**Comment by lsytj0413** [comment]
> LGTM! Many thanks! @lsytj0413 wanna take a look?

There is a lot of enhancement in this PR, include many `UnmarshalJSON` implement. I'd prefer opening an issue for every fix  
& improment to describe why we need this. That would help us understand why we need this.

File: model/action.go [hunk 3]
```diff
@@ -50,14 +50,16 @@ type actionForUnmarshal Action
...
 // UnmarshalJSON implements json.Unmarshaler
 func (a *Action) UnmarshalJSON(data []byte) error {
+
 	v := actionForUnmarshal{
 		ActionDataFilter: ActionDataFilter{UseResults: true},
 	}
 	err := json.Unmarshal(data, &v)
 	if err != nil {
-		return err
+		return fmt.Errorf("action value '%s' is not supported, it must be an object or string", string(data))
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/action.go:59]: a better message is returned than this one.

File: model/event_state.go [hunk 4]
```diff
@@ -36,12 +38,18 @@ type eventStateForUnmarshal EventState
 
 // UnmarshalJSON unmarshal EventState object from json bytes
 func (e *EventState) UnmarshalJSON(data []byte) error {
+	var timeout EventStateTimeout
+	if err := json.Unmarshal(data, &timeout); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/event_state.go:44]: to not get nil pointer :)

**Comment by spolti** [comment]
> > LGTM! Many thanks! @lsytj0413 wanna take a look?
> 
> There is a lot of enhancement in this PR, include many `UnmarshalJSON` implement. I'd prefer opening an issue for every fix & improment to describe why we need this. That would help us understand more about it.

Not really, the main goal is to fix the nil pointers. If you want to break it in different issues, I am not against but I don't see a strong reason to do it.

**Comment by lsytj0413** [comment]
> Not really, the main goal is to fix the nil pointers. If you want to break it in different issues, I am not against but I don't see a strong reason to do it.

There is a lot change of files & lines, an issue still helpful.

File: parser/testdata/workflows/greetings-v08-spec.sw.yaml [hunk 5]
```diff
@@ -35,19 +106,64 @@ states:
...
           refName: sendTextFunction
           arguments:
             message: "${ .singlemessage }"
+        eventRef:
+          triggerEventRef: example1
+          resultEventRef: example1
+          # Added "resultEventTimeout" for action eventref
+          resultEventTimeout: PT12H
+    actionExecTimeout: PT11H
+    stateExecTimeout:
```
[the remark this answers is not in the document]
Comment (spolti) [reply parser/testdata/workflows/greetings-v08-spec.sw.yaml:136]: good point, thanks.

File: model/event_state.go [hunk 6]
```diff
@@ -80,7 +88,35 @@ func (o *OnEvents) UnmarshalJSON(data []byte) error {
...
+	data = bytes.TrimSpace(data)
+	if len(data) == 0 {
+		// TODO: Normalize error messages
+		return fmt.Errorf("no bytes to unmarshal")
+	}
+
+	var err error
+	switch data[0] {
+	case '"':
+		e.StateExecTimeout.Total, err = unmarshalString(data)
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/event_state.go:108]: indeed, not needed.

File: model/event.go [hunk 7]
```diff
@@ -46,6 +47,17 @@ func EventStructLevelValidation(structLevel validator.StructLevel) {
 	}
 }
 
+// EventRefStructLevelValidation custom validator for event kind consumed
+func EventRefStructLevelValidation(structLevel validator.StructLevel) {
+	eventRef := structLevel.Current().Interface().(EventRef)
+	if len(eventRef.ResultEventTimeout) > 0 {
+		err := val.ValidateISO8601TimeDuration(eventRef.ResultEventTimeout)
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/event.go:54]: I see, yes, this field already contains this tags, this custom validator is not needed :)
thanks.

File: model/delay_state.go [hunk 8]
```diff
@@ -16,15 +16,24 @@ package model
...
-	val.GetValidator().RegisterStructValidation(
-		DelayStateStructLevelValidation,
-		DelayState{},
-	)
+	val.GetValidator().RegisterStructValidation(DelayStateStructLevelValidation, DelayState{})
+}
+
+// DelayStateStructLevelValidation custom validator for DelayState Struct
+func DelayStateStructLevelValidation(structLevel validator.StructLevel) {
+	delayState := structLevel.Current().Interface().(DelayState)
```
[the remark this answers is not in the document]
Comment (spolti) [reply model/delay_state.go:30]: +1

**Comment by spolti** [comment]
@ricardozanini @lsytj0413 can you guys please take another look?

created this issue then: https://github.com/serverlessworkflow/sdk-go/issues/107

**Review by ricardozanini — COMMENTED** [review]
@spolti I think there's only one comment left to deal before merging this. `timeouts` is optional, we should return `nil`.

File: model/action.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lsytj0413) [reply model/action.go:59]: When the unmarshal return err, data could be an object byte array. This error message unable to describe specific error information in detail.

File: model/sleep_state.go [hunk 9]
```diff
@@ -28,3 +33,24 @@ type SleepState struct {
 type SleepStateTimeout struct {
 	StateExecTimeout StateExecTimeout `json:"stateExecTimeout,omitempty"`
```
Comment (lsytj0413) [inline model/sleep_state.go:34]: If the StateExecTimeout is optional, we should change it's type to *StateExecTimeout？

File: model/sleep_state.go [hunk 10]
```diff
@@ -28,3 +33,24 @@ type SleepState struct {
...
 }
+
+type sleepStateForUnmarshal SleepState
+
+func (s *SleepState) UnmarshalJSON(data []byte) error {
```
Comment (lsytj0413) [inline model/sleep_state.go:39]: It's seems useless, can you add a unittest to describe why we need this?

File: parser/testdata/workflows/greetings-v08-spec.sw.yaml [hunk 11]
```diff
@@ -14,17 +14,91 @@
...
+    type: switch
+    eventConditions:
+      - eventRef: visaApprovedEvent
+        transition: HandleApprovedVisa
+      - eventRef: visaRejectedEvent
+        transition: HandleRejectedVisa
+    eventTimeout: PT1H
+    stateExecTimeout:
+      total: PT1S
+      single: PT2S
```
Comment (lsytj0413) [inline parser/testdata/workflows/greetings-v08-spec.sw.yaml:87]: eventTimeout & stateExecTimeout is under `timeouts` property like other state.

File: model/event_state.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (spolti) [reply model/event_state.go:44]: agreed.

File: model/sleep_state.go (hunk 9 above)
Comment (spolti) [reply model/sleep_state.go:34]: the
```
Timeouts *SleepStateTimeout
```

already is a pointer, is there a need to use pointers again here?

File: model/sleep_state.go (hunk 10 above)
Comment (spolti) [reply model/sleep_state.go:39]: indeed, not needed.

File: parser/testdata/workflows/greetings-v08-spec.sw.yaml (hunk 11 above)
Comment (spolti) [reply parser/testdata/workflows/greetings-v08-spec.sw.yaml:87]: yep, copied from others json examples from the tests ones, seems that they were wrong as well.
thanks for the heads up.

**Comment by spolti** [comment]
small fixes for https://github.com/serverlessworkflow/sdk-go/issues/70

**Review by ricardozanini — APPROVED** [review]

File: model/states.go [hunk 12]
```diff
@@ -137,10 +166,32 @@ type OperationState struct {
...
+// UnmarshalJSON unmarshal EventState object from json bytes
+func (o *OperationState) UnmarshalJSON(data []byte) error {
+
+	var timeout OperationStateTimeout
+	if err := json.Unmarshal(data, &timeout); err != nil {
+		return err
+	}
+
+	v := operationStateForUnmarshal{
+		Timeouts: &timeout,
```
Comment (lsytj0413) [inline model/states.go:180]: we should keep it nil when the user didn't define it.

File: model/states.go [hunk 13]
```diff
@@ -171,9 +243,30 @@ type InjectState struct {
...
+
+// UnmarshalJSON unmarshal EventState object from json bytes
+func (i *InjectState) UnmarshalJSON(data []byte) error {
+	var timeout InjectStateTimeout
+	if err := json.Unmarshal(data, &timeout); err != nil {
+		return err
+	}
+
+	v := injectStateForUnmarshal{
+		Timeouts: &timeout,
```
Comment (lsytj0413) [inline model/states.go:256]: we should keep it nil when the user didn't define it.

File: model/states.go (hunk 12 above)
Comment (spolti) [reply model/states.go:180]: indeed

**Pull request merged**